pkg/agent/endpoints/sds/handler.go

```diff
@@ -0,0 +1,314 @@
+	auth_v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2/auth"
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	discovery_v2 "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2"
+	"github.com/gogo/protobuf/types"
+	"github.com/sirupsen/logrus"
+	attestor "github.com/spiffe/spire/pkg/agent/attestor/workload"
+	"github.com/spiffe/spire/pkg/agent/manager/cache"
+	"github.com/spiffe/spire/pkg/common/auth"
+	"github.com/spiffe/spire/pkg/common/bundleutil"
+	"github.com/spiffe/spire/pkg/common/pemutil"
+	"github.com/spiffe/spire/pkg/common/telemetry"
+	"github.com/spiffe/spire/proto/common"
+	"github.com/zeebo/errs"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	sdsAPI = "sds_api"
+	sdsPID = "sds_pid"
+)
+
+type Manager interface {
+	SubscribeToCacheChanges(key cache.Selectors) cache.Subscriber
+	FetchWorkloadUpdate(selectors []*common.Selector) *cache.WorkloadUpdate
+}
+
+type HandlerConfig struct {
+	Attestor attestor.Attestor
+	Manager  Manager
+	Metrics  telemetry.Metrics
+	Log      logrus.FieldLogger
+}
+
+type Handler struct {
+	c HandlerConfig
+
+	hooks struct {
+		// test hook used to synchronize receipt of a stream request
+		received chan struct{}
+	}
+}
+
+func NewHandler(config HandlerConfig) *Handler {
+	return &Handler{c: config}
+}
+
+func (h *Handler) StreamSecrets(stream discovery_v2.SecretDiscoveryService_StreamSecretsServer) error {
+	_, selectors, metrics, done, err := h.startCall(stream.Context())
+	if err != nil {
+		return err
+	}
+	defer done()
+
+	metrics = metrics
+
+	sub := h.c.Manager.SubscribeToCacheChanges(selectors)
+	defer sub.Finish()
+
+	updch := sub.Updates()
+	reqch := make(chan *api_v2.DiscoveryRequest, 1)
+	errch := make(chan error, 1)
+
+	go func() {
+		for {
+			req, err := stream.Recv()
+			if err != nil {
+				if status.Code(err) == codes.Canceled || err == io.EOF {
+					err = nil
+				}
+				errch <- err
+				return
+			}
+			reqch <- req
+		}
+	}()
+
+	var versionInfo string
+	var lastNonce string
+	var upd *cache.WorkloadUpdate
+	var req *api_v2.DiscoveryRequest
+	for {
+		select {
+		case newReq := <-reqch:
+			h.c.Log.Debugf("stream-secrets: recv: res=%q ver=%q typ=%q nonce=%q", newReq.ResourceNames, newReq.VersionInfo, newReq.TypeUrl, newReq.ResponseNonce)
+			h.triggerReceivedHook()
+
+			// The nonce should be empty (if we've never sent a response) or
+			// match the last sent nonce, otherwise the request should be
+			// ignored.
+			if ((lastNonce == "") != (newReq.ResponseNonce == "")) || lastNonce != newReq.ResponseNonce {
+				h.c.Log.Warnf("unexpected nonce %q (expected %q); ignoring request", newReq.ResponseNonce, lastNonce)
+				continue
+			}
+
+			req = newReq
+
+			if upd == nil {
+				// Workload update has not been received yet.
+				continue
+			}
+			if req.ErrorDetail != nil {
+				// The caller has failed to apply the secrets. Wait until the
+				// next update to send down new info.
```

Comment:
Ahh... ok yes, I see that now. I misinterpreted one of the error diagrams in that doc.

I guess I figured that there are probably transient error conditions on envoy side, but maybe there are not...? My main concern here was a transient envoy error preventing an SVID update, then subsequently falling offline because presumably that SVID would expire before we try to create a new one (which would really be the third one).

I am definitely no expert here... I guess we can take it as is and see if we run into anything